components/asic/common.c

```diff
@@ -24,4 +34,95 @@ int _largest_power_of_two(int num)
     }
 
     return 1 << power;
+}
+
+int count_asic_chips(uint16_t asic_count, uint16_t chip_id)
+{
+    uint8_t buffer[11] = {0};
+
+    int chip_counter = 0;
+    while (true) {
+        int received = SERIAL_rx(buffer, 11, 1000);
+        if (received == 0) break;
+
+        if (received == -1) {
+            ESP_LOGW(TAG, "Error reading CHIP_ID");
+            break;
+        }
+
+        if (received != 11) {
+            ESP_LOGW(TAG, "Invalid CHIP_ID response length");
+            ESP_LOG_BUFFER_HEX(TAG, buffer, received);
+            break;
+        }
+
+        uint16_t received_preamble = (buffer[0] << 8) | buffer[1];
+        if (received_preamble != PREAMBLE) {
+            ESP_LOGW(TAG, "Preamble mismatch: got 0x%04x, expected 0x%04x", received_preamble, PREAMBLE);
+            ESP_LOG_BUFFER_HEX(TAG, buffer, received);
+            continue;
+        }
+
+        uint16_t received_chip_id = (buffer[2] << 8) | buffer[3];
+        if (received_chip_id != chip_id) {
+            ESP_LOGW(TAG, "CHIP_ID response mismatch: got 0x%04x, expected 0x%04x", received_chip_id, chip_id);
+            ESP_LOG_BUFFER_HEX(TAG, buffer, received);
+            continue;
+        }
+
+        if (crc5(buffer + 2, received - 2) != 0) {
+            ESP_LOGW(TAG, "Checksum failed on CHIP_ID response");
+            ESP_LOG_BUFFER_HEX(TAG, buffer, received);
+            continue;
+        }
+
+        ESP_LOGI(TAG, "Chip %d detected: CORE_NUM: 0x%02x ADDR: 0x%02x", chip_counter, buffer[4], buffer[5]);
+
+        chip_counter++;
+    }    
+    
+    if (chip_counter != asic_count) {
+        ESP_LOGW(TAG, "%i chip(s) detected on the chain, expected %i", chip_counter, asic_count);
+    }
+
+    return chip_counter;
+}
+
+esp_err_t receive_work(uint8_t * buffer, int buffer_size)
+{
+    int received = SERIAL_rx(buffer, buffer_size, 10000);
+
+    if (received < 0) {
+        ESP_LOGE(TAG, "UART error in serial RX");
+        return ESP_FAIL;
+    }
+
+    if (received == 0) {
+        ESP_LOGE(TAG, "UART timeout in serial RX");
```

Comment:
`asic_timeout_counter` in the old receive_work never increased to 2, as it is local scope. Now it occasionally logs an error, so I don't know if we should increase the timeout here, or just ignore a timeout (e.g. no logging).